packages/cmrjs/cmr-utils.js

```diff
@@ -476,83 +529,145 @@ function getCreds() {
+ */
+function buildMergedEchoURLObject(URLlist = [], originalURLlist = [], removedURLs = [],
+  URLTypes, URLlistFieldFilter) {
+  let filteredURLObjectList = URLlist.filter((urlObj) => URLTypes.includes(urlObj.Type));
+  filteredURLObjectList = filteredURLObjectList.map((urlObj) => omit(urlObj, URLlistFieldFilter));
+  return mergeURLs(originalURLlist, filteredURLObjectList, removedURLs);
+}
 
 /**
  * After files are moved, this function creates new online access URLs and then updates
  * the S3 ECHO10 CMR XML file with this information.
  *
- * @param {Object} cmrFile - cmr xml file object to be updated
- * @param {Array<Object>} files - array of file objects
- * @param {string} distEndpoint - distribution endpoint from config
- * @param {BucketsConfig} buckets - stack BucketConfig instance
+ * @param {Object} params - parameter object
+ * @param {Object} params.cmrFile - cmr xml file object to be updated
+ * @param {Array<Object>} params.files - array of file objects
+ * @param {string} params.distEndpoint - distribution endpoint from config
+ * @param {BucketsConfig} params.buckets - stack BucketConfig instance
  * @returns {Promise} returns promised updated metadata object.
  */
-async function updateEcho10XMLMetadata(cmrFile, files, distEndpoint, buckets) {
-  let newURLs = constructOnlineAccessUrls({ files, distEndpoint, buckets });
-  newURLs = newURLs.map((urlObj) => omit(urlObj, ['Type', 'Description']));
-  const removedURLs = onlineAccessURLsToRemove(files, buckets);
-
+async function updateEcho10XMLMetadata({
+  cmrFile,
+  files,
+  distEndpoint,
+  buckets,
+  s3CredsEndpoint = 's3credentials'
+}) {
   // add/replace the OnlineAccessUrls
   const filename = getS3UrlOfFile(cmrFile);
   const metadataObject = await metadataObjectFromCMRXMLFile(filename);
   const metadataGranule = metadataObject.Granule;
-
   const updatedGranule = { ...metadataGranule };
-  let originalURLs = _get(metadataGranule, 'OnlineAccessURLs.OnlineAccessURL', []);
 
-  // If there is only one OnlineAccessURL in the file, it comes back as an object and not an array
-  if (!Array.isArray(originalURLs)) {
-    originalURLs = [originalURLs];
-  }
+  const originalOnlineAccessURLs = [].concat(_get(metadataGranule,
+    'OnlineAccessURLs.OnlineAccessURL', []));
+  const originalOnlineResourceURLs = [].concat(_get(metadataGranule,
+    'OnlineResources.OnlineResource', []));
+  const originalAssociatedBrowseURLs = [].concat(_get(metadataGranule,
+    'AssociatedBrowseImageUrls.ProviderBrowseUrl', []));
 
-  const mergedURLs = mergeURLs(originalURLs, newURLs, removedURLs);
-  _set(updatedGranule, 'OnlineAccessURLs.OnlineAccessURL', mergedURLs);
-  metadataObject.Granule = updatedGranule;
+  const removedURLs = onlineAccessURLsToRemove(files, buckets);
+  const newURLs = constructOnlineAccessUrls({ files, distEndpoint, buckets })
+    .concat(getS3CredentialsObject(urljoin(distEndpoint, s3CredsEndpoint)));
```

Comment:
This line and the previous don't really read right to me. First of all, I don't like `.concat` as a solution because it's not really extensible - if someone wants to add another baked-in `relatedURL`, this solution doesn't really help. One thing to note, though, is that only URLs we want baked in, like the `s3credentials` url, would be added here.

Second, `constructOnlineAccessUrls` doesn't feel like a good name for that function anymore. It's not _only_ constructing an `OnlineAccessUrl` object. Something like `constructCMRURLObject` rings more true for me.